**<h1>Business Understanding</h1>**

**<h3> Problem </h3>**

The problem that we aim to solve is the accurate forecasting of a movie's success based on various features like cast, crew, budget, revenue, release dates, and more. This solution is significant because it allows movie studios and production companies to make informed decisions, optimize investments, mitigate risks, and enhance audience satisfaction by predicting a movie's success accurately. Solving this problem is crucial as it enables efficient resource management, better project selection, and improved marketing strategies within the entertainment industry. By contributing a predictive model that leverages data-driven insights, this approach can drive greater profitability and audience engagement. The results obtained from this predictive model can be viably used to guide decision-making processes related to movie production, marketing strategies, and resource allocation in the entertainment industry

**<h3> Dataset </h3>**

We are using a dataset from Kaggle titled "Millions of Movies." It is an excellent resource for movie analysis and prediction due to its comprehensive coverage and dynamic nature. It encompasses about a million movies, spanning from the 1900s to the present day, providing a rich historical context and a wide variety of data points for analysis. The dataset includes detailed information on cast, crew, plot keywords, budget, revenue, posters, release dates, and languages, which are crucial factors in predicting a movie's success. Additionally, its daily updates and enrichment with TMDB API ensure that the dataset remains current and reflective of the latest trends and releases in the movie industry. This level of detail and breadth makes it a valuable asset for building accurate and robust predictive models that can help studios and production companies make informed decisions

The dataset includes a blend of nominal and numeric features, such as movie ID, title, language, budget, languages, genres, production company, revenue and detailed information on cast and crew. It encompass a wide range of movies, providing a rich source of data for analysis. The inclusion of both quantitative data (e.g., budget, revenue) and qualitative data (e.g., plot keywords, genres) allows for a comprehensive approach to predicting movie success. This combination of breadth, depth, and timeliness makes it a valuable resource for developing a predictive model of movie success, as it captures a wide array of factors that can influence a movie's performance in the market.

Link to the dataset: <link>https://www.kaggle.com/datasets/akshaypawar7/millions-of-movies</link>

**<h3> Proposed Solution </h3>**

The proposed solution is the development of a machine learning-based model to predict the success of movies before their release. This solution is significant because it addresses the challenge of uncertainty in the movie industry by providing a data-driven approach to forecast profitability, which is crucial for making informed investment decisions. It can help studios and production companies to optimize their investments, reduce financial risks, and tailor marketing strategies to enhance audience satisfaction. By accurately predicting movie success, the model can contribute to the bigger picture by improving the efficiency of resource allocation, project selection, and marketing efforts in the entertainment industry. The results of the model can be used to guide decision-making processes at various stages of movie production and distribution

**<h1>Data Understanding and Exploration</h1>**

**<h3>1. Dataset Filtering</h3>**

The first step in the data understanding phase is to check for the availabilty of the target feature. Based on the proposed solution, the target feature should accurately measure the success of the movie. For that, we decided to derive our target feature which will be the ratio of the revenue and budget. Consequently, if the ratio obtained is greater than 2, meaning if the movie makes twice as much as its net cost of production, it would be considered a success. We plan to utilize 3 levels of success indicators low, medium and high as per the ratio.

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('movies.csv',sep=r',', skipinitialspace=True)
df.head(20)

id                                title  \
0   615656                    Meg 2: The Trench   
1   758323                  The Pope's Exorcist   
2   667538     Transformers: Rise of the Beasts   
3   640146    Ant-Man and the Wasp: Quantumania   
4   677179                            Creed III   
5   385687                               Fast X   
6   614479              Insidious: The Red Door   
7   646389                                Plane   
8   569094  Spider-Man: Across the Spider-Verse   
9   956101                    The Eighth Clause   
10  572802         Aquaman and the Lost Kingdom   
11  447277                   The Little Mermaid   
12  594767             Shazam! Fury of the Gods   
13  455476                Knights of the Zodiac   
14  753342                             Napoleon   
15  933131                      Badland Hunters   
16  968051                           The Nun II   
17  965839                  Lord of the Streets   
18  670292                          The Creator   
19  772515              Huesera: The Bone Woman   

                                        genres original_language  \
0                Action-Science Fiction-Horror                en   
1                      Horror-Mystery-Thriller                en   
2             Action-Adventure-Science Fiction                en   
3             Action-Adventure-Science Fiction                en   
4                                 Drama-Action                en   
5                        Action-Crime-Thriller                en   
6                      Horror-Mystery-Thriller                en   
7                    Action-Adventure-Thriller                en   
8   Action-Adventure-Animation-Science Fiction                en   
9                                     Thriller                la   
10                    Action-Adventure-Fantasy                en   
11            Adventure-Family-Fantasy-Romance                en   
12             Action-Comedy-Fantasy-Adventure                en   
13                    Fantasy-Action-Adventure                en   
14                           History-War-Drama                en   
15                Action-Science Fiction-Drama                ko   
16                     Horror-Mystery-Thriller                en   
17                                      Action                en   
18             Science Fiction-Action-Thriller                en   
19                        Drama-Horror-Mystery                es   

                                             overview  popularity  \
0   An exploratory dive into the deepest depths of...    8763.998   
1   Father Gabriele Amorth Chief Exorcist of the V...    5953.227   
2   When a new threat capable of destroying the en...    5409.104   
3   Super-Hero partners Scott Lang and Hope van Dy...    4425.387   
4   After dominating the boxing world Adonis Creed...    3994.342   
5   Over many missions and against impossible odds...    3583.861   
6   To put their demons to rest once and for all J...    3512.648   
7   After a heroic job of successfully landing his...    2618.646   
8   After reuniting with Gwen Stacy Brooklyn’s ful...    2550.738   
9   Kat and Borja appear to be a perfect couple bu...    2259.303   
10  Black Manta still driven by the need to avenge...    2141.691   
11  The youngest of King Triton’s daughters and th...    2118.742   
12  Billy Batson and his foster siblings who trans...    2010.980   
13  When a headstrong street orphan Seiya in searc...    1882.774   
14  An epic that details the checkered rise and fa...    1811.360   
15  After a deadly earthquake turns Seoul into a l...    1704.650   
16  In 1956 France a priest is violently murdered ...    1696.620   
17  When Jason Dyson refuses to make his prized fi...    1691.825   
18  Amid a future war between the human race and t...    1595.817   
19  Valeria's joy at becoming a first-time mother ...    1436.445   

                                 production_companies rel

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 722480 entries, 0 to 722479
Data columns (total 20 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   id                    722480 non-null  int64  
 1   title                 722474 non-null  object 
 2   genres                511978 non-null  object 
 3   original_language     722480 non-null  object 
 4   overview              604113 non-null  object 
 5   popularity            722480 non-null  float64
 6   production_companies  337256 non-null  object 
 7   release_date          670592 non-null  object 
 8   budget                722480 non-null  float64
 9   revenue               722480 non-null  float64
 10  runtime               688110 non-null  float64
 11  status                722480 non-null  object 
 12  tagline               108313 non-null  object 
 13  vote_average          722480 non-null  float64
 14  vote_count            722480 non-null  float64
 15  

In [4]:
num_zeros = ((df['budget'] != 0) & (df['revenue'] != 0)).sum()
num_zeros

11190

We observe that the budget and revenue parameters combined, are only present for a total of 11190 records out of the total. Hence, we proceed with further data exploration on the filtered dataset where both the parameters are present.

In [6]:
nonzero_df = df[(df['budget'] != 0) & (df['revenue'] != 0)]

#saving the filtered dataset into a csv file
nonzero_df.to_csv('movies_filtered.csv', index=False)

In [7]:
#deleting the unfiltered dataframe and loading the newly created one
del df
df = pd.read_csv('movies_filtered.csv',sep=r',', skipinitialspace=True)
df.head()

id                              title  \
0  615656                  Meg 2: The Trench   
1  758323                The Pope's Exorcist   
2  667538   Transformers: Rise of the Beasts   
3  640146  Ant-Man and the Wasp: Quantumania   
4  677179                          Creed III   

                             genres original_language  \
0     Action-Science Fiction-Horror                en   
1           Horror-Mystery-Thriller                en   
2  Action-Adventure-Science Fiction                en   
3  Action-Adventure-Science Fiction                en   
4                      Drama-Action                en   

                                            overview  popularity  \
0  An exploratory dive into the deepest depths of...    8763.998   
1  Father Gabriele Amorth Chief Exorcist of the V...    5953.227   
2  When a new threat capable of destroying the en...    5409.104   
3  Super-Hero partners Scott Lang and Hope van Dy...    4425.387   
4  After dominating the boxing world Adonis Creed...    3994.342   

                                production_companies release_date  \
0  Apelles Entertainment-Warner Bros. Pictures-di...   2023-08-02   
1  Screen Gems-2.0 Entertainment-Jesus & Mary-Wor...   2023-04-05   
2  Skydance-Paramount-di Bonaventura Pictures-Bay...   2023-06-06   
3             Marvel Studios-Kevin Feige Productions   2023-02-15   
4  Metro-Goldwyn-Mayer-Proximity Media-Balboa Pro...   2023-03-01   

        budget      revenue  runtime    status  \
0  129000000.0  352056482.0    116.0  Released   
1   18000000.0   65675816.0    103.0  Released   
2  200000000.0  407045464.0    127.0  Released   
3  200000000.0  475766228.0    125.0  Released   
4   75000000.0  269000000.0    116.0  Released   

                                             tagline  vote_average  \
0                                  Back for seconds.         7.079   
1  Inspired by the actual files of Father Gabriel...         7.433   
2                                     Unite or fall.         7.340   
3            Witness the beginning of a new dynasty.         6.507   
4                      You can't run from your past.         7.262   

   vote_count                                            credits  \
0      1365.0  Jason Statham-Wu Jing-Shuya Sophia Cai-Sergio ...   
1       545.0  Russell Crowe-Daniel Zovatto-Alex Essoe-Franco...   
2      1007.0  Anthony Ramos-Dominique Fishback-Luna Lauren V...   
3      2811.0  Paul Rudd-Evangeline Lilly-Jonathan Majors-Kat...   
4      1129.0  Michael B. Jordan-Tessa Thompson-Jonathan Majo...   

                                            keywords  \
0                based on novel or book-sequel-kaiju   
1  spain-rome italy-vatican-pope-pig-possession-c...   
2  peru-alien-end of the world-based on cartoon-b...   
3  hero-ant-sequel-superhero-based on comic-famil...   
4  philadelphia pennsylvania-husband wife relatio...   

                        poster_path                     backdrop_path  \
0  /4m1Au3YkjqsxF8iwQy0fPYSxE0h.jpg  /qlxy8yo5bcgUw2KAmmojUKp4rHd.jpg   
1  /9JBEPLTPSm0d1mbEcLxULjJq9Eh.jpg  /hiHGRbyTcbZoLsYYkO4QiCLYe34.jpg   
2  /gPbM0MK8CP8A174rmUwGsADNYKD.jpg  /woJbg7ZqidhpvqFGGMRhWQNoxwa.jpg   
3  /qnqGbB22YJ7dSs4o6M7exTpNxPz.jpg  /m8JTwHFwX7I7JY5fPe4SjqejWag.jpg   
4  /cvsXj3I9Q2iyyIo95AecSd1tad7.jpg  /5i6SjyDbDWqyun8klUuCxrlFbyw.jpg   

                                     recommendations  
0  1006462-298618-569094-1061181-346698-1076487-6...  
1  713704-296271-502356-1076605-1084225-1008005-9...  
2  496450-569094-298618-385687-877100-598331-4628...  
3  823999-676841-868759-734048-267805-965839-1033...  
4  965839-267805-943822-842942-1035806-823999-107...

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11190 entries, 0 to 11189
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    11190 non-null  int64  
 1   title                 11190 non-null  object 
 2   genres                10692 non-null  object 
 3   original_language     11190 non-null  object 
 4   overview              10820 non-null  object 
 5   popularity            11190 non-null  float64
 6   production_companies  10060 non-null  object 
 7   release_date          10747 non-null  object 
 8   budget                11190 non-null  float64
 9   revenue               11190 non-null  float64
 10  runtime               11143 non-null  float64
 11  status                11190 non-null  object 
 12  tagline               8449 non-null   object 
 13  vote_average          11190 non-null  float64
 14  vote_count            11190 non-null  float64
 15  credits            

In [9]:
df['revenue_budget_ratio'] = df['revenue'] / df['budget']
#Checking central tendencies for the derived feature
mean_ratio = df['revenue_budget_ratio'].mean()
median_ratio = df['revenue_budget_ratio'].median()
range_ratio = df['revenue_budget_ratio'].max() - df['revenue_budget_ratio'].min()

print("Mean of revenue_budget_ratio:", mean_ratio)
print("Median of revenue_budget_ratio:", median_ratio)
print("Range of revenue_budget_ratio:", range_ratio)

Mean of revenue_budget_ratio: 25066.423088188672
Median of revenue_budget_ratio: 1.75
Range of revenue_budget_ratio: 100004216.0


The mean is evidently quite high than the median. The central tendencies are breaking probably because of outliers.

In [10]:
s_df = df.sort_values(by='revenue_budget_ratio', ascending=False)
s_df.head()


id                   title  genres original_language  \
10260  932428      Żyraf: Czas Zemsty  Horror                pl   
10291  888201        GRU VS BOBUX MAN     NaN                en   
10006  698480        Amizade Desfeita     NaN                en   
9970   737626        A Isca e o Anzol     NaN                pt   
10689  535928  Wooden Slippers (2007)     NaN                tl   

                                                overview  popularity  \
10260                                                NaN         0.6   
10291                                                NaN         0.6   
10006  Blaire Mitch Jess Adam Ken and Val are in a ch...         0.6   
9970                                                 NaN         0.6   
10689                             The best movie of 2007         0.6   

      production_companies release_date  budget      revenue  ...    status  \
10260                  NaN          NaN     1.0  100004215.0  ...  Released   
10291                  NaN          NaN     1.0   90000000.0  ...  Released   
10006                  NaN          NaN     1.0   64056643.0  ...  Released   
9970                   NaN          NaN     1.0    9999999.0  ...  Released   
10689                  NaN          NaN   100.0  500000000.0  ...  Released   

                tagline vote_average  vote_count  \
10260               NaN          8.0         1.0   
10291               NaN         10.0         1.0   
10006               NaN          0.0         0.0   
9970                NaN          0.0         0.0   
10689  Oh nang gewa mu?         10.0         1.0   

                                                 credits keywords  \
10260  Fabian Żygadło-Antek Kaminski-Kacper Podczasia...      NaN   
10291                                  kakav420-kraccbak      NaN   
10006                                        Blaire lily      NaN   
9970                                                 NaN      NaN   
10689                                                NaN      NaN   

                            poster_path                     backdrop_path  \
10260  /1oGhKAwgzcQtOsA5nXZDzo7dhUL.jpg  /iyIxtg6EFEtKI3wNEhXs9ppP1Zb.jpg   
10291                               NaN  /goC4BL2ae4YYTP8SHd3Yo6jIdgd.jpg   
10006                               NaN                               NaN   
9970                                NaN                               NaN   
10689  /ejtnZGYX8NnQUHFMsPWqPtliFff.jpg                               NaN   

      recommendations revenue_budget_ratio  
10260             NaN          100004215.0  
10291             NaN           90000000.0  
10006             NaN           64056643.0  
9970              NaN            9999999.0  
10689             NaN            5000000.0  

[5 rows x 21 columns]

**<h3>Identifying the Data Scales, Data Types and Data domain for the features</h3>**

In [11]:
for column in df.columns:
    unique_values = df[column].unique()
    print(f"Unique values in {column}: {unique_values}")
    

Unique values in id: [615656 758323 667538 ... 512178 506977 501079]
Unique values in title: ['Meg 2: The Trench' "The Pope's Exorcist"
 'Transformers: Rise of the Beasts' ...
 '10 INSANE World Records OF ALL TIME' '100W Bulb' 'Chanel']
Unique values in genres: ['Action-Science Fiction-Horror' 'Horror-Mystery-Thriller'
 'Action-Adventure-Science Fiction' ... 'Thriller-Animation'
 'Animation-Fantasy-Comedy' 'Drama-Fantasy-Comedy']
Unique values in original_language: ['en' 'fi' 'ja' 'fr' 'es' 'zh' 'cn' 'hi' 'ko' 'te' 'ru' 'pt' 'it' 'de'
 'th' 'no' 'id' 'ar' 'da' 'nb' 'sv' 'tl' 'cs' 'el' 'pl' 'fa' 'uk' 'kn'
 'ro' 'he' 'ms' 'tr' 'ta' 'nl' 'hu' 'ml' 'vi' 'et' 'ca' 'af' 'ur' 'gl'
 'ku' 'bn' 'iu' 'sr' 'mr' 'sk' 'bm' 'pa' 'ne' 'gu' 'lt' 'az' 'si' 'km'
 'dz' 'xx' 'jv' 'ka' 'bg' 'am' 'ab' 'ht' 'dv' 'bs' 'sq' 'or' 'mn' 'lv'
 'lg' 'mo']
Unique values in overview: ['An exploratory dive into the deepest depths of the ocean of a daring research team spirals into chaos when a malevolent mining operati

In [12]:

features_df = pd.DataFrame(columns=['col_name', 'desc', 'data_type', 'scale(N,O,I,R)', 'domain', 'missing_values'])

# Iterate through columns of the original DataFrame and populate the column information DataFrame
for col in df.columns:
    col_info = {
        'col_name': col,
        'desc': '',  # Add description if available
        'data_type': df[col].dtype,
        'scale(N,O,I,R)': '',  
        'domain': df[col].unique(),  # Add domain if available
        'missing_values': df[col].isnull().sum()
    }
    features_df = pd.concat([features_df, pd.DataFrame([col_info])], ignore_index=True)
    
#Setting the data types
features_df.loc[features_df['data_type'] == 'object', 'data_type'] = 'String'
features_df.loc[features_df['col_name'] == 'release_date', 'data_type'] = 'Date'
features_df

#Setting the data scale
## ---------------data scale col---------------------------
scales = ['N', 'N', 'N', 'N', 'N', 'I', 'N', 'I', 'R', 'R', 'R', 'N', 'N', 'R', 'R', 'N', 'N', 'N', 'N', 'N','R']
features_df['scale(N,O,I,R)'] = scales
# N, O, I, R are Nominal, Ordinal, Interval, Ratio


#Setting the descriptions
## ---------------description col---------------------------
col_descriptions = {
    'id': 'Unique identifier for the movie',
    'title': 'Title of the movie',
    'genres': 'Genres associated with the movie',
    'original_language': 'Original language of the movie',
    'overview': 'Brief description of the movie',
    'popularity': 'TMDB lifetime popularity score',
    'production_companies': 'Companies involved in the production',
    'release_date': 'Date the movie was released',
    'budget': 'Amount of money spent to produce the movie',
    'revenue': 'Amount of money the movie generated',
    'runtime': 'Duration of the movie in minutes',
    'status': 'Current status of the movie (e.g., released, in production)',
    'tagline': 'Short phrase that captures the essence of the movie',
    'vote_average': 'Average user rating of the movie',
    'vote_count': 'Number of user ratings for the movie',
    'credits': 'Information about the cast and crew of the movie',
    'keywords': 'Keywords associated with the movie',
    'poster_path': 'File path for the movie poster image',
    'backdrop_path': 'File path for the movie backdrop image',
    'recommendations': 'Other movies recommended based on this one',
    'revenue_budget_ratio': 'Ratio of the movie\'s revenue to its budget'
}
features_df['desc'] = features_df['col_name'].map(col_descriptions)
features_df

col_name                                               desc  \
0                     id                    Unique identifier for the movie   
1                  title                                 Title of the movie   
2                 genres                   Genres associated with the movie   
3      original_language                     Original language of the movie   
4               overview                     Brief description of the movie   
5             popularity                     TMDB lifetime popularity score   
6   production_companies               Companies involved in the production   
7           release_date                        Date the movie was released   
8                 budget         Amount of money spent to produce the movie   
9                revenue                Amount of money the movie generated   
10               runtime                   Duration of the movie in minutes   
11                status  Current status of the movie (e.g., released, i...   
12               tagline  Short phrase that captures the essence of the ...   
13          vote_average                   Average user rating of the movie   
14            vote_count               Number of user ratings for the movie   
15               credits   Information about the cast and crew of the movie   
16              keywords                 Keywords associated with the movie   
17           poster_path               File path for the movie poster image   
18         backdrop_path             File path for the movie backdrop image   
19       recommendations         Other movies recommended based on this one   
20  revenue_budget_ratio         Ratio of the movie's revenue to its budget   

   data_type scale(N,O,I,R)  \
0      int64              N   
1     String              N   
2     String              N   
3     String              N   
4     String              N   
5    float64              I   
6     String              N   
7       Date              I   
8    float64              R   
9    float64              R   
10   float64              R   
11    String              N   
12    String              N   
13   float64              R   
14   float64              R   
15    String              N   
16    String              N   
17    String              N   
18    String              N   
19    String              N   
20   float64              R   

                                               domain missing_values  
0   [615656, 758323, 667538, 640146, 677179, 38568...              0  
1   [Meg 2: The Trench, The Pope's Exorcist, Trans...              0  
2   [Action-Science Fiction-Horror, Horror-Mystery...            498  
3   [en, fi, ja, fr, es, zh, cn, hi, ko, te, ru, p...              0  
4   [An exploratory dive into the deepest depths o...            370  
5   [8763.998, 5953.227, 5409.104, 4425.387, 3994....              0  
6   [Apelles Entertainment-Warner Bros. Pictures-d...           1130  
7   [2023-08-02, 2023-04-05, 2023-06-06, 2023-02-1...            443  
8   [129000000.0, 18000000.0, 200000000.0, 7500000...              0  
9   [352056482.0, 65675816.0, 407045464.0, 4757662...              0  
10  [116.0, 103.0, 127.0, 125.0, 142.0, 107.0, 140...             47  
11  [Released, Post Production, In Production, Rum...              0  
12  [Back for seconds., Inspired by the actual fil...           2741  
13  [7.079, 7.433, 7.34, 6.507, 7.262, 7.275, 6.75...              0  
14  [1365.0, 545.0, 1007.0, 2811.0, 1129.0, 3694.0...              0  
15  [Jason Statham-Wu Jing-Shuya Sophia Cai-Sergio...            437  
16  [based on novel or book-sequel-kaiju, spain-ro...           1890  
17  [/4m1Au3YkjqsxF8iwQy0fPYSxE0h.jpg, /9JBEPLTPSm...            544  
18  [/qlxy8yo5bcgUw2KAmmojUKp4rHd.jpg, /hiHGRbyTcb...           1625  
19  [1006462-298618-569094-1061181-346698-1076487-...           4007  
20  [2.729120015503876, 3.6486564444444443, 2.0352...              0